File: frontend/campus_app/lib/services/likeService.dart
import 'package:dio/dio.dart';
import 'package:social_media/models/data_response_message.dart';
import 'package:social_media/models/response_message.dart';
import 'package:social_media/models/story_dto.dart';
import 'package:social_media/models/post_dto.dart';
import 'package:social_media/models/search_account_dto.dart';

class LikeService {
  late Dio _dio;
  
  // Constructor'da Dio alabilecek şekilde düzenle
  LikeService([Dio? dio]) {
    if (dio != null) {
      _dio = dio;
    } else {
      // Dio verilmezse boş bir instance oluştur, initDio() ile güncellenecek
      _dio = Dio();
    }
  }
  
  // Service'i kullanmadan önce çağrılmalı
  Future<void> initDio(String token) async {
    _dio.options.headers['Authorization'] = 'Bearer $token';
  }
  
  // Hikayeyi beğenme
  Future<ResponseMessage> likeStory(String token, String storyId) async {
    await initDio(token);
    final response = await _dio.post(
      '/likes/story/$storyId',
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Gönderiyi beğenme
  Future<ResponseMessage> likePost(String token, String postId) async {
    await initDio(token);
    final response = await _dio.post(
      '/likes/post/$postId',
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Hikaye beğenisini kaldırma (Unlike)
  Future<ResponseMessage> unlikeStory(String token, String storyId) async {
    await initDio(token);
    final response = await _dio.delete(
      '/likes/story/$storyId',
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Gönderi beğenisini kaldırma (Unlike)
  Future<ResponseMessage> unlikePost(String token, String postId) async {
    await initDio(token);
    final response = await _dio.delete(
      '/likes/post/$postId',
    );
    return ResponseMessage.fromJson(response.data);
  }

  // Beğendiği hikayeleri listele
  Future<DataResponseMessage<List<StoryDTO>>> getUserLikedStories(String token) async {
    await initDio(token);
    final response = await _dio.get(
      '/likes/stories',
    );
    return DataResponseMessage<List<StoryDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => StoryDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Beğendiği gönderileri listele
  Future<DataResponseMessage<List<PostDTO>>> getUserLikedPosts(String token, {int page = 0, int size = 10}) async {
    try {
      await initDio(token);
      final response = await _dio.get(
        '/likes/posts',
        queryParameters: {
          'page': page,
          'size': size
        },
      );
      
      if (response.data['success'] == true) {
        final List<dynamic> postList = response.data['data'] ?? [];
        try {
          final posts = postList.map((item) => PostDTO.fromJson(item)).toList();
          return DataResponseMessage<List<PostDTO>>(
            message: response.data['message'] ?? 'Başarılı',
            data: posts,
            isSuccess: true,
          );
        } catch (e) {
          print('PostDTO dönüştürme hatası: $e');
          return DataResponseMessage<List<PostDTO>>(
            message: 'Veri dönüştürme hatası: $e',
            data: [],
            isSuccess: false,
          );
        }
      } else {
        return DataResponseMessage<List<PostDTO>>(
          message: response.data['message'] ?? 'Bir hata oluştu',
          data: [],
          isSuccess: false,
        );
      }
    } catch (e) {
      print('LikeService getUserLikedPosts error: $e');
      return DataResponseMessage<List<PostDTO>>(
        message: 'Bir hata oluştu: $e',
        data: [],
        isSuccess: false,
      );
    }
  }

  // Belirli bir tarihten sonra gönderiye yapılan beğenileri getir
  Future<DataResponseMessage<List<PostDTO>>> getPostLikesAfter(String token, String postId, String dateTime) async {
    await initDio(token);
    final response = await _dio.get(
      '/likes/post/$postId/likes-after/$dateTime',
    );
    return DataResponseMessage<List<PostDTO>>.fromJson(
      response.data,
      (data) => (data as List).map((item) => PostDTO.fromJson(item as Map<String, dynamic>)).toList(),
    );
  }

  // Belirtilen hikayede belirli bir kullanıcının beğenisini arama
  Future<DataResponseMessage<SearchAccountDTO>> searchUserInStoryLikes(String token, String storyId, String username) async {
    await initDio(token);
    final response = await _dio.get(
      '/likes/story/$storyId/search/$username',
    );
    return DataResponseMessage<SearchAccountDTO>.fromJson(
      response.data,
      (data) => SearchAccountDTO.fromJson(data as Map<String, dynamic>),
    );
  }

  // Belirtilen gönderide belirli bir kullanıcının beğenisini arama
  Future<DataResponseMessage<SearchAccountDTO>> searchUserInPostLikes(String token, String postId, String username) async {
    await initDio(token);
    final response = await _dio.get(
      '/likes/post/$postId/search/$username',
    );
    return DataResponseMessage<SearchAccountDTO>.fromJson(
      response.data,
      (data) => SearchAccountDTO.fromJson(data as Map<String, dynamic>),
    );
  }
  
  Future<bool> checkPostLike(String token, String postId) async {
    await initDio(token);
    final response = await _dio.get(
      '/likes/post/$postId/check',
    );
    return response.data['liked'] as bool;
  }
  
  // Post beğenme/beğeni kaldırma
  Future<ResponseMessage> toggleLike(String token, String postId) async {
    await initDio(token);
    final response = await _dio.post(
      '/likes/post/$postId/toggle',
    );
    return ResponseMessage.fromJson(response.data);
  }



}

File: backend/Bingol-University-Social-Media-main/src/main/java/bingol/campus/like/business/abstracts/LikeService.java
package bingol.campus.like.business.abstracts;

import bingol.campus.followRelation.core.exceptions.BlockingBetweenStudent;
import bingol.campus.like.core.exceptions.AlreadyLikedException;
import bingol.campus.like.core.exceptions.PostNotFoundLikeException;
import bingol.campus.like.core.exceptions.StoryNotFoundLikeException;
import bingol.campus.post.core.exceptions.PostNotFoundException;
import bingol.campus.post.core.exceptions.PostNotIsActiveException;
import bingol.campus.post.core.response.PostDTO;
import bingol.campus.response.DataResponseMessage;
import bingol.campus.response.ResponseMessage;
import bingol.campus.story.core.exceptions.NotFollowingException;
import bingol.campus.story.core.exceptions.StoryNotActiveException;
import bingol.campus.story.core.exceptions.StoryNotFoundException;
import bingol.campus.story.core.exceptions.StudentProfilePrivateException;
import bingol.campus.story.core.response.StoryDTO;
import bingol.campus.student.core.response.SearchAccountDTO;
import bingol.campus.student.exceptions.StudentNotFoundException;

import java.util.List;
import java.util.UUID;

public interface LikeService {
    ResponseMessage likeStory(String username, UUID storyId) throws StoryNotFoundException, StudentNotFoundException, StoryNotActiveException, BlockingBetweenStudent, NotFollowingException, AlreadyLikedException, StudentProfilePrivateException;

    ResponseMessage likePost(String username, UUID postId) throws StudentNotFoundException, PostNotFoundException, PostNotIsActiveException, BlockingBetweenStudent, NotFollowingException, AlreadyLikedException, StudentProfilePrivateException;

    ResponseMessage unlikeStory(String username, UUID storyId) throws StoryNotFoundException, StudentNotFoundException, StoryNotFoundLikeException;

    ResponseMessage unlikePost(String username, UUID postId) throws StudentNotFoundException, PostNotFoundException, PostNotFoundLikeException;

    DataResponseMessage<List<StoryDTO>> getUserLikedStories(String username) throws StudentNotFoundException;

    DataResponseMessage<List<PostDTO>> getUserLikedPosts(String username) throws StudentNotFoundException;

    DataResponseMessage<List<PostDTO>> getPostLikesAfter(UUID postId, String dateTime) throws PostNotFoundException;


    DataResponseMessage<SearchAccountDTO> searchUserInPostLikes(String username, UUID postId, String username1) throws PostNotFoundException, StudentNotFoundException, NotFollowingException, BlockingBetweenStudent, StudentProfilePrivateException;

    DataResponseMessage<SearchAccountDTO> searchUserInStoryLikes(String username, UUID storyId, String username1) throws StudentNotFoundException, StoryNotFoundException, NotFollowingException, BlockingBetweenStudent, StudentProfilePrivateException;

    boolean checkPostLike(String username, UUID postId) throws StudentNotFoundException;

    ResponseMessage togglePostLike(String username, UUID postId) throws StudentNotFoundException, PostNotFoundException;
}
